Fixes #3362 Profiles Integration

tadean opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This PR aims to provide an integration with the profiles service, for `az_person` import. It adds a new experimental module, `az_person_profile_import` that offers a migration from pulling information from the profiles API.

## Related issues
#3362 

## How to test

- **NOTE: to test this PR, you must have an api key for the profiles API.**
- enable `az_person_profiles_import`
- If you intend to import a large number of people, it is very recommend to enable `az_http` for performance reasons
- Visit `/admin/content/profiles/import`
- Check that the message appears, `You must first configure a Profiles API token here.`
- Visit `/admin/config/az-quickstart/settings/az-person-profiles-import`
- Add your profiles API token to the **API Token** field
- return to `/admin/content/profiles/import`
- Enter some NetIDs. Enter one NetID that does not exist
- For `Choose how profiles are imported`, select `All listed profiles`
- Click Import
- Check for status messages indicating which people were imported. Verify that a warning message is output for the netID that did not exist
- Examine imported content for accuracy
- For imported people, verify that fields from the profiles API cannot be edited, and a warning  message is displayed that the person was imported

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [x] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

tadean commented:
**Todo:**

- [ ] Determine namespacing
- [ ] Add additional fields

trackleft commented on modules/custom/az_person/az_person_profile_import/migrations/az_person_profile_import.yml:
```diff
@@ -0,0 +1,101 @@
+    source: titles
+    process:
+      value: desc
+  field_az_degrees:
+    -
+      plugin: sub_process
+      source: degrees
+      include_source: true,
+      source_key: original_source
+      process:
+        0: 'properties/degree_title'
+        1:
+          -
+            plugin: callback
+            callable: array_column
+            unpack_source: true
+            source:
+              - institution
+              - 'original_source/constants/NAME'
+          -
+            plugin: concat
+            delimiter: ', '
+    -
+      plugin: sub_process
+      process:
+        value:
+          -
+            plugin: concat
+            source:
+              - 0
+              - 1
+            delimiter: ', '
```
After some research on a profile integration on a departmental site, I found they were importing degrees so that they were formatted like:

```html
<ul>
  <li>Bachelor of Science, Computer Science, University of Example, 2020</li>
</ul>

```

The migration appears to fomats degrees like:

```html
<ul>
  <li>Bachelor of Science, University of Example</li>
</ul>

```

Not sure  if it helps, but here is the logic from the departmental importer.

Degrees are sorted by year

```
  private function degree_sort_cmp($a, $b)
  {
    $a_degree_conferred_year = intval(isset($a->CONFERRED_DEGREE->year) ? $a->CONFERRED_DEGREE->year : 0);
    $b_degree_conferred_year = intval(isset($b->CONFERRED_DEGREE->year) ? $b->CONFERRED_DEGREE->year : 0);
    if ($a_degree_conferred_year == $b_degree_conferred_year)
    {
      return 0;
    }
    return ($a_degree_conferred_year < $b_degree_conferred_year) ? 1 : -1;
  }

```
Degrees are formatted
```
  public function getDegrees()
  {
    $html = "";
    if (!empty($this->degrees))
    {
      $html .= "<ul>";
      foreach ($this->degrees as $degree) {
        $html .= "<li>" . $degree->properties->degree_type;
        if (isset($degree->properties->discipline)) {
          $html .= ", " . $degree->properties->discipline;
        }
        if (isset($degree->institution[0]->name)) {
          $html .= ", " . $degree->institution[0]->name;
        }
        if (isset($degree->CONFERRED_DEGREE->year)) {
          $html .= ", " . $degree->CONFERRED_DEGREE->year;
        }
        $html .= "</li>";

      }
      $html .= "</ul>";
    }
    return $html;
  }

```

trackleft replied:
Here is what the degrees field looks like for reference
<img width="1088" alt="image" src="https://github.com/az-digital/az_quickstart/assets/1023167/1b11529d-1bd1-4ea7-a330-b248160e1106">

tadean replied:
Agreed, I think we should add this.

trackleft commented on modules/custom/az_person/az_person_profile_import/az_person_profile_import.module:
```diff
@@ -0,0 +1,23 @@
+
+/**
+ * Implements hook_entity_presave().
+ */
```
I wonder if this could be done in an Event Subscriber, on the [onPreRowSave event](https://api.drupal.org/api/drupal/core%21modules%21migrate%21src%21Event%21MigrateEvents.php/constant/MigrateEvents%3A%3APRE_ROW_SAVE/10)

tadean replied:
I'm not sure if this is possible currently. On testing this out, the migration `OnPreRowSave` appears to only have access to the migration row, not the entity, for access to the revision log settings. it also seems like the Entity side of Drupal currently [lacks the equivalent entity events for subscribers to interact with](https://www.drupal.org/project/drupal/issues/2551893). That seems surprising...

It seems like an issue in progress also for entity migrations to interact with revisions in a more direct way: [https://www.drupal.org/project/drupal/issues/2988382](https://www.drupal.org/project/drupal/issues/2988382)

trackleft commented on modules/custom/az_person/az_person_profile_import/migrations/az_person_profile_import.yml:
```diff
@@ -0,0 +1,101 @@
+      selector: 'Degrees'
+
+process:
+  type:
```
This should be possible now.
```suggestion
  field_az_netid: netid
  type:
```

tadean replied:
Good idea, I think it would be good to add this.

tadean commented:
- Added netid
- Added additional preprocessing and ordering of degrees

trackleft commented on modules/custom/az_person/az_person_profile_import/migrations/az_person_profile_import.yml:
```diff
@@ -0,0 +1,76 @@
+      selector: 'Degrees'
+
+process:
+  type:
```
Do you think this is a good idea?
```suggestion
  revision_log:
    plugin: default_value
    default_value: 'Imported from Profiles'
  type:
```

trackleft commented on modules/custom/az_person/az_person_profile_import/migrations/az_person_profile_import.yml:
```diff
@@ -0,0 +1,76 @@
+      selector: 'Degrees'
+
+process:
+  type:
+    plugin: default_value
+    default_value: az_person
```
```suggestion
  type:
    plugin: default_value
    default_value: az_person
  revision_log:
    plugin: default_value
    default_value: 'Imported from Profiles'
```

trackleft replied:
I think we should mention the specific fields it is updating.

tadean replied:
> Do you think this is a good idea?

Migrate imports don't seem to currently create revisions: [Add option to create new revision in entity migration](https://www.drupal.org/project/drupal/issues/2988382). We are currently forcing this to happen in `hook_entity_presave()`

tadean replied:
> I think we should mention the specific fields it is updating.

@trackleft Do you think that the revision log message should mention exactly which fields were changed?

tadean commented:
- Added optional `nid` mapping to existing nodes using `entity_lookup`
- Added `field_az_person_imported` boolean for tracking imported persons
- Added edit form hook that disables certain fields for imported persons

tadean commented:
Removed `field_az_person_imported` and implemented @joeparsons suggestion to use the migrate map to determine whether the person has been imported or not. There are some circumstances where this data could be removed, but after considering it, it seems preferable to use this as the source of truth rather than adding additional fields to the content type, given that other additional import types will be coming in the future, and these fields create overhead that may not be used on every Quickstart site.

tadean commented:
Added functionality that mentions specific field changes to revision log message.

tadean commented:
- [ ] Need decision from data stewards of profiles API on if Address field should be imported

tadean commented:
> * [x]  Need decision from data stewards of profiles API on if Address field should be imported

Profiles API data stewards have approved the use of any publicly visible profile information.

tadean commented:
- [ ] Address Field
- [ ] Recurring imports TBD
- [ ] Data stewards request to show where to get support on profiles data:

> Regarding support burden, colleges using imported profile information should have a website function for contacting local support.

tadean commented:
We discussed the following potential namespacing for the proposed import modules

- `az_person_profiles_import`
- `az_person_eds_import`

tadean commented:
Resolved an issue @joeparsons pointed out that after the rename of the module namespace, the folder itself hadn't been changed to match.

danahertzberg commented:
Follow up improvement: "Choose how profiles are imported" selection option text and help text

danahertzberg commented:
Better to test locally rather than on probo to keep our public/live API keys hidden.

bberndt-uaz commented on modules/custom/az_person/az_person_profiles_import/src/Form/AZPersonProfilesImportForm.php:
```diff
@@ -0,0 +1,141 @@
+      '#type' => 'select',
+      '#title' => $this->t('Choose how profiles are imported'),
+      '#options' => [
+        'normal' => $this->t('New profiles'),
+        'update' => $this->t('All listed profiles'),
+        'track_changes' => $this->t('Profiles that have been updated since last import'),
```
Here is my suggestion for rewording the import options to possibly provide a little more clarity for end users. I also switched the last two so that the options are ordered in terms of how many records will be imported.

```suggestion
        'normal' => $this->t('Import new profiles only'),
        'track_changes' => $this->t('Import new profiles and profiles updated since the last import'),
        'update' => $this->t('Import all listed profiles'),
```
